src/scippnexus/nxobject.py

```
@@ -466,14 +466,19 @@ def __getattr__(self, attr: str) -> Union[Any, 'NXobject']:
                                   f"to obtain all matches instead of obj.{attr}.")
 
     def __dir__(self):
-        nxclasses = []
-        for val in self.values():
-            if isinstance(val, NXobject):
-                nxclasses.append(val.nx_class)
         keys = super().__dir__()
+        try:
+            nxclasses = []
+            for val in self.values():
+                if isinstance(val, NXobject):
+                    nxclasses.append(val.nx_class)
+        except Exception:
```

Can this be made more concrete? E.g. `(RuntimeError, KeyError)`?

---

I don't know, `values()` instantiates the various specific `NXobject` subclasses. We don't know for sure what can go wrong there?

---

But are those failures reasons to omit the attr or would it be a hard failure that should raise an exception with the user?

---

Neither: It most likely indicates a problem about an unrelated child, we should probably try to avoid `values()` here?

---

@jl-wynen I refactored this, please have a look at the latest commit.